### PR TITLE
feat: add profiler util

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

## Description

- Add a simple profiler utility
- Add config settings to enable profiling graphs, output dir for the profiles, and a prefix for the profile outputs
- Add helper script to generate graph visualizations from the profiles
- Add `snakeviz` and `gprof2dot` as dev deps (you also need graphviz but it's not available as a standalone python package, comments added to link to it)
- Use profile in invocation processor (only if profiling is enabled)
- Add tests for profiler

## QA Instructions, Screenshots, Recordings

- `pip install -e '.[dev]'` to get `snakeviz` and `gprof2dot`
- install [graphviz](https://www.graphviz.org/download/) 
- In `invokeai.yaml`, under `Development:` add `profile_graphs: true` (profiles will be written to `$INVOKEAI_ROOT/profiles` by default, can change this with `profiles_dir: /some/path/`)
- Generate
- Run `snakeviz /path/to/profiles/session_id.prof` to explore as a flamegraph
- Run `./scripts/generate_profile_graphs.sh /path/to/profiles` to generate PDFs visualizations of the graphs as a directed graph, will do this for all the graphs. Output SVG or PNG instead by adding second arg `svg` or `png`. PDF is nice bc its a very small file, and you can ctrl+f to find function names

## Merge Plan

This PR can be merged when approved

### CODE REVIEW

**Review by hipsterusername — APPROVED** [review]

File: invokeai/app/util/profiler.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    Usage
+    ```
+      # Create a profiler
+      profiler = Profiler(logger, output_dir)
+      # Start a new profile
+      profiler.new("my_profile")
+      profiler.enable()
+      # Do stuff
+      profiler.disable()
+      profiler.dump()
```
Comment (RyanJDick) [inline invokeai/app/util/profiler.py:21]: Why not just have two methods? I.e. `profiler.start("my_profile")` and `profiler.stop()`.

File: invokeai/app/util/profiler.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    ```
+    """
+
+    def __init__(self, logger: Logger, output_dir: Path, prefix: Optional[str] = None) -> None:
+        self.logger = logger
+        self.output_dir = output_dir
+        self.output_dir.mkdir(parents=True, exist_ok=True)
+        self.profiler: Optional[cProfile.Profile] = None
+        self.profile_id: Optional[str] = None
+        self.prefix = prefix
```
Comment (RyanJDick) [inline invokeai/app/util/profiler.py:44]: I think the intended public interface could be made more clear by marking private fields with a leading underscore. Something like:

```suggestion
        self._logger = logger
        self._output_dir = output_dir
        self._output_dir.mkdir(parents=True, exist_ok=True)
        self._profiler: Optional[cProfile.Profile] = None
        self._prefix = prefix
        
        self.profile_id: Optional[str] = None
```

**Review by RyanJDick — COMMENTED** [review]
Nice, this will be convenient. I left a couple of small suggestions.

File: invokeai/app/util/profiler.py (hunk 1 above)
Comment (psychedelicious) [reply invokeai/app/util/profiler.py:21]: I was trying to retain the flexibility and API of `cProfile.Profile`, but it's definitely less ergonomic.

Also, in the structured stats PR, I added an arg to this dump method to optionally write the stats to JSON. The intention was to output all the profiling data at once, but it couples the stats service to this profiler, which is not great. The two classes shouldn't interact like that.

I agree that just a start and stop on the profiler is better. Can add a method to the stats service to write its data to disk, and just call that at the same time as we write the profiler's data.

**Pull request merged**